### PR TITLE
storage controller: send startup compute notifications in background

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

Previously, we try to send compute notifications in startup_reconcile before completing that function, with a time limit.  Any notifications that don't happen within the time limit result in tenants having their `pending_compute_notification` flag set, which causes them to spawn a Reconciler next time the background reconciler loop runs.

This causes two problems:
- Spawning a lot of reconcilers after startup caused a spike in memory (this is addressed in https://github.com/neondatabase/neon/pull/7493)
- After https://github.com/neondatabase/neon/pull/7493, spawning lots of reconcilers will block some other operations, e.g. a tenant creation might fail due to lack of reconciler semaphore units while the controller is busy running all the Reconcilers for its startup compute notifications.

When the code was first written, ComputeHook didn't have internal ordering logic to ensure that notifications for a shard were sent in the right order.  Since that was added in https://github.com/neondatabase/neon/pull/7088, we can use it to avoid waiting for notifications to complete in startup_reconcile.

Related to: https://github.com/neondatabase/neon/issues/7460

## Summary of changes

- Add a `notify_background` method to ComputeHook.
- Call this from startup_reconcile instead of doing notifications inline
- Process completions from `notify_background` in `process_results`, and if a notification failed then set the `pending_compute_notification` flag on the shard.

The result is that we will only spawn lots of Reconcilers if the compute notifications _fail_, not just because they take some significant amount of time.

Test coverage for this case is in https://github.com/neondatabase/neon/pull/7475

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: storage_controller/src/compute_hook.rs [hunk 1]
```diff
@@ -508,6 +499,94 @@ impl ComputeHook {
...
+        self: &Arc<Self>,
+        notifications: Vec<(TenantShardId, NodeId, ShardStripeSize)>,
+        result_tx: tokio::sync::mpsc::Sender<Result<(), (TenantShardId, NotifyError)>>,
+        cancel: &CancellationToken,
+    ) {
+        let mut maybe_sends = Vec::new();
+        for (tenant_shard_id, node_id, stripe_size) in notifications {
+            let maybe_send_result = self.notify_prepare(tenant_shard_id, node_id, stripe_size);
+            maybe_sends.push((tenant_shard_id, maybe_send_result))
+        }
```
Comment (petuhovskiy) [inline storage_controller/src/compute_hook.rs:516]: Here we're taking a send_lock for each `ComputeHookTenant` and these locks are held potentially for a long time (until `stream` is fully processed).

It means that `ComputeHook::notify` calls can be blocked waiting for these locks. Probably not an issue at the current scale, but worth mentioning.

**Review by petuhovskiy — APPROVED** [review]
LGTM. Looked through the storage-controller code, now more or less understand how it works

**Comment by petuhovskiy** [comment]
Giving it another thought, I think  `ComputeHook` service can just have a single unbounded channel for accepting requests and it doesn't need to have `notify()`, `notify_background()` or any other pub functions. For notifying about errors, it can have a single outgoing channel that will be processed in `process_results`.

**Comment by jcsp** [comment]
> Giving it another thought, I think ComputeHook service can just have a single unbounded channel for accepting requests and it doesn't need to have notify(), notify_background() or any other pub functions. For notifying about errors, it can have a single outgoing channel that will be processed in process_results.

The important characteristics of notify() are:
- is that it shouldn't complete until the notification has been processed -- when we do something like a migration, we must not detach the old location until we know the compute has been notified of the new location.
- one tenant's notification does not block another tenants (e.g. if we had a global channel for notifications, we couldn't pause processing it while we wait for one tenant's notifications to complete).

Those behaviors could indeed still be implemented on top of channels though.

**Pull request merged**